onnxruntime/core/optimizer/propagate_cast_ops.cc

```diff
@@ -0,0 +1,709 @@
+      if (producer && IsCastTo(producer, TensorProto::FLOAT)) {
+        casts.push_back(producer);
+        require_type_change.insert(input);
+        continue;
+      }
+      all_float_inputs_have_casts = false;
+      break;
+    }
+    if (has_float_inputs && all_float_inputs_have_casts && casts.size() > 1) {
+      VLOGS(logger, 1) << "PropagateFP32CastsFromInputsToOutputs: Removed Cast nodes "
+                       << ConcatNames<std::vector<Node*>>(casts)
+                       << " feeding the same compute node " << node->Name();
+      for (Node* cast : casts) {
+        RemoveCastNodes(graph, {cast}, removed_nodes);
+      }
+      std::unordered_set<NodeArg*> node_args;
+      for (NodeArg* output : node->MutableOutputDefs()) {
+        if (output->Exists() && IsType(*output, TensorProto::FLOAT)) {
+          node_args.insert(output);
+        }
+      }
+      InsertCastNodes(graph, node_args, false, removed_nodes);
+      ChangeTypeToFP16(require_type_change, logger);
+      VLOGS(logger, 1) << "PropagateFP32CastsFromInputsToOutputs: Inserted Cast node to "
+                       << ConcatNames(node_args);
+      modified = true;
+    }
+  }
+  return modified;
+}
+
+// PropagateFP16CastsFromOutputsToInputs
+// This non recursive fusion, checks whether the given node is fp16 safe op and
+// whether all floatingpoint outputs are cast to fp16
+// and propagates cast op to the floatingpoint inputs.
+static bool PropagateFP16CastsFromOutputsToInputs(Graph& graph, Node* node,
+                                                  std::deque<onnxruntime::NodeIndex>& removed_nodes,
+                                                  size_t level,
+                                                  const logging::Logger& logger) {
+  bool modified = false;
+  if (IsFP16Allow(node->OpType(), level)) {
+    bool has_float_outputs = false;
+    bool all_float_outputs_have_casts = true;
+    std::vector<Node*> casts;  // Cast nodes to propagate.
+    std::vector<NodeArg*>& outputs = node->MutableOutputDefs();
+    std::unordered_set<NodeArg*> require_type_change;
+    for (auto iter = outputs.begin(); iter != outputs.end() && all_float_outputs_have_casts; ++iter) {
+      NodeArg* output = *iter;
+      if (!IsType(*output, TensorProto::FLOAT)) {
+        continue;
+      }
+      has_float_outputs = true;
+      std::vector<Node*> consumers = graph.GetMutableConsumerNodes(output->Name());
+      for (auto node_iter = consumers.begin(); node_iter != consumers.end() && all_float_outputs_have_casts; ++node_iter) {
+        Node* consumer = *node_iter;
+        if (consumer != nullptr &&
+            std::find(removed_nodes.begin(), removed_nodes.end(), consumer->Index()) == removed_nodes.end() &&
+            IsCastTo(consumer, TensorProto::FLOAT16)) {
+          casts.push_back(consumer);
+          continue;
+        }
+        all_float_outputs_have_casts = false;
+      }
+      require_type_change.insert(output);
+    }
```

Comment:
similarlly, can this be replaced with graph_utils::FindChildrenByType(*node, "Cast") ?

Reply:
Same as above.